File: 139.单词拆分.cpp
/*
 * @lc app=leetcode.cn id=139 lang=cpp
 *
 * [139] 单词拆分
 */

// @lc code=start
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> set;
        for (auto word: wordDict) {
            set.insert(word);
        }
        vector<int> index(1, -1);
        for (int i = 0; i < s.size(); i++) {
            for (auto ind: index) {
                if (set.count(s.substr(ind+1, i-ind))) {
                    index.push_back(i);
                    break;
                }
            }
        }
        return index.back() == s.size()-1;
    }
};
// @lc code=end

File: 139.单词拆分.py
# @before-stub-for-debug-begin
from python3problem139 import *
from typing import *
# @before-stub-for-debug-end

#
# @lc app=leetcode.cn id=139 lang=python3
#
# [139] 单词拆分
#

# @lc code=start
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        dic = set(wordDict)
        dp = [False]*(len(s)+1)
        dp[0] = True
        for i in range(1, len(s)+1):
            for j in range(i-1, -1, -1):
                dp[i] = dp[i] or dp[j] and (s[j:i] in dic)
        return dp[len(s)]
# @lc code=end

